src/trace-api.ts

```diff
@@ -175,49 +154,58 @@ export class StackdriverTracer implements Tracer {
-    const incomingTraceContext: IncomingTraceContext = {options: 1};
-    let parsedContext: util.TraceContext|null = null;
-    if (isString(options.traceContext) &&
-        this.config!.contextHeaderBehavior !==
-            TraceContextHeaderBehavior.IGNORE) {
-      parsedContext = util.parseContextFromHeader(options.traceContext);
-    }
-    if (parsedContext) {
-      if (parsedContext.options === undefined) {
-        // If there are no incoming option flags, default to 0x1.
-        parsedContext.options = 1;
+    const parseContext = (stringifiedTraceContext?: string|null) => {
+      const parsedContext = isString(stringifiedTraceContext) ?
+          util.parseContextFromHeader(stringifiedTraceContext) :
+          null;
+      if (parsedContext) {
+        if (parsedContext.options === undefined) {
+          // If there are no incoming option flags, default to 0x1.
+          parsedContext.options = 1;
+        }
       }
-      Object.assign(incomingTraceContext, parsedContext);
-    } else if (
-        this.config!.contextHeaderBehavior ===
-        TraceContextHeaderBehavior.REQUIRE) {
-      incomingTraceContext.options = 0;
-    }
+      return parsedContext as Required<util.TraceContext>| null;
+    };
+    const traceContext = parseContext(options.traceContext);
 
     // Consult the trace policy.
-    const locallyAllowed = this.policy!.shouldTrace({
+    const shouldTrace = this.policy!.shouldTrace({
       timestamp: Date.now(),
       url: options.url || '',
-      method: options.method || ''
+      method: options.method || '',
+      traceContext,
+      options
     });
-    const remotelyAllowed = !!(
-        incomingTraceContext.options & Constants.TRACE_OPTIONS_TRACE_ENABLED);
 
     let rootContext: RootSpan&RootContext;
+
     // Don't create a root span if the trace policy disallows it.
-    if (!locallyAllowed || !remotelyAllowed) {
+    if (!shouldTrace) {
       rootContext = UNTRACED_ROOT_SPAN;
     } else {
       // Create a new root span, and invoke fn with it.
-      const traceId =
-          incomingTraceContext.traceId || (uuid.v4().split('-').join(''));
-      const parentId = incomingTraceContext.spanId || '0';
       const name = this.config!.rootSpanNameOverride(options.name);
-      rootContext = new RootSpanData(
-          {projectId: '', traceId, spans: []}, /* Trace object */
-          name,                                /* Span name */
-          parentId,                            /* Parent's span ID */
-          options.skipFrames || 0);
+      if (traceContext && this.config!.propagateTraceContextFromHeader) {
+        rootContext = new RootSpanData(
+            {
+              projectId: '',
+              traceId: traceContext.traceId,
+              spans: []
+            },                   /* Trace object */
+            name,                /* Span name */
+            traceContext.spanId, /* Parent's span ID */
+            options.skipFrames || 0);
```

Comment:
It is hard to see what is different between the `then`/`else` blocks. Instead use local variables like so:
```ts
const traceId = traceContext && this.config!.propagateTraceContextFromHeader ?
  traceContext.traceId || traceId: uuid.v4().split('-').join('');
const spanId = traceContext && this.config!.propagateTraceContextFromHeader ?
  traceContext.spanId : 0;
```

Reply:
I've removed the then/else blocks.